### PR TITLE
Improved warning message when userscript is not mode 755.

### PR DESCRIPTION
**Pull request by Gemba**
Refers to: https://retropie.org.uk/forum/topic/29651/emulation-station-game-start-script-not-executable

### CODE REVIEW

**Comment by cmitu** [comment]
Shouldn't we rather fix the execution bit detection instead of requiring a specific file mode ? If the file is executable (u/g/a),  the detection shouldn't fail.
I see no mention in the original scripting PR about such restriction (script being just `u+x`).

**Comment by Gemba** [comment]
Maybe it's just my pov. ES is running as user pi, thus the scripts should only be writeable by this user. However, I can go with executable test only for ugo, -- Scripting is the only module using `isExecutable()`, so no side effects except the lowered hardening, but the scripting part has never ment to be rock hardened/secured in the first place.

**Comment by cmitu** [comment]
I think it's ok to allow `g=+x` and `o=+x` (without checking for `g=+w` or `o=+w`).

**Comment by cmitu** [comment]
After thinking about it a bit more, I think the `0755` is not be the intended file mode for the script to be considered in events. 

Since scripts can also be distributed in multiple places that includes `$INSTDIR/resources/scripts`, where scripts will be `root` owned. If set to just mode `0755`, those scripts would not be executable by the running user and pretty much useless.

**Comment by Gemba** [comment]
Would be check for `7__` ok? Thus `u=rwx` mandatory, `g/o` bits can be set as needed but won't fail this test.

**Comment by cmitu** [comment]
`7__` might not be enough - if the script is installed as `root`, but it's not executable for `o`, the check would pass, but execution would fail. 
Outside of checking every possible combination to ensure the ES user can execute effectively execute the script/program, I think the `isExecutable` should check if any `x` bit is set and fail (with the error message) when none is set.

**Comment by Gemba** [comment]
Did some more testing. "any `x`" is too lax. When called from ES on Rpi minimum is `0o100` for binaries and `0o500` for scripts. Without `u=[r]x` the syscall fails with permission denied. Group/Other `[r]x` flags can be present if needed but user exec/read permission is mandatory.

**Comment by cmitu** [comment]
Can't say I can reproduce this. 
I have a global script in `/opt/retropie/supplementary/emulationstation/scripts/game-start` with permissions `o+x` and it's executed fine. If the script is owned by the running user, then yes, `07__` check is ok, but if it's not, then it's not enough.
```
$ pwd
/opt/retropie/supplementary/emulationstation-dev/scripts/game-start
$ ls -l 
-rw-r--r-x 1 root root 71 Feb 19 18:49 system.sh
```

**Comment by Gemba** [comment]
Ah, I see. Thanks for the info. I adapted the PR to at least on exec-bit on `u/g/o` (pre-check) and a hint if the permissions are insufficient for the account/user Emulationstation runs with to execute the file in question.

**Comment by cmitu** [comment]
Looks fine, except the info/error messages - I find it too confusing. What about just  _<script> is not executable by the current user_ for the _ENOENT_ error and _<script> is not executable_ for the first ?

**Comment by Gemba** [comment]
Thanks for reviewing. And yes, agreed: I edited it to more terse log messages.

**Comment by theofficialgman** [comment]
@jrassa I think this commit is ready for a merge

File: es-core/src/utils/FileSystemUtil.cpp [hunk 1]
```diff
@@ -740,16 +740,17 @@ namespace Utils
...
+			const mode_t mask = S_IFREG;
+			const mode_t mask_exec = S_IXUSR | S_IXGRP | S_IXOTH;
 			struct stat64 info;
 
 			// check if stat64 succeeded
 			if(stat64(path.c_str(), &info) != 0)
 				return false;
 
-			// check for mask attributes only
-			return ((info.st_mode & mask) == info.st_mode);
```
Comment (tomaz82) [inline es-core/src/utils/FileSystemUtil.cpp:752]: Wouldn't the old code be enough if just changing it to compare to mask rather than info.st_mode?

Comment (Gemba) [reply es-core/src/utils/FileSystemUtil.cpp:752]: Changing to `(info.st_mode & mask) == mask` in the current implementation would match 755 and "higher", but not "lower" like 451 or 700 or 011. So this PR suggests to check only for at least one exec-bit (hence the two checks: one for regular files one for at least one exec bit). Execution can still fail but then the it is the user's responsibility (eg. mounting the script folder `noexec`). Rationale to allow also "lower" permissions: Compiled binaries need only a `x`-bit, interpreted scripts need `rx`.

**Comment by tomaz82** [comment]
Thanks!

**Pull request merged**